zrml/neo-swaps/src/benchmarking.rs

```diff
@@ -125,86 +292,147 @@ mod benchmarks {
 
     #[benchmark]
     fn sell() {
-        let alice: T::AccountId = whitelisted_caller();
-        let base_asset = Asset::Ztg;
+        let alice = whitelisted_caller();
         let market_id =
-            create_market_and_deploy_pool::<T>(alice, base_asset, 2u16, _10.saturated_into());
+            create_market_and_deploy_pool::<T>(alice, Asset::Ztg, 2u16, _10.saturated_into());
         let asset_in = Asset::CategoricalOutcome(market_id, 0);
         let amount_in = _1.saturated_into();
         let min_amount_out = 0u8.saturated_into();
 
-        let bob: T::AccountId = whitelisted_caller();
+        let helper = BenchmarkHelper::<T>::new();
+        let bob = helper.accounts().next().unwrap();
         assert_ok!(T::MultiCurrency::deposit(asset_in, &bob, amount_in));
 
         #[extrinsic_call]
         _(RawOrigin::Signed(bob), market_id, 2, asset_in, amount_in, min_amount_out);
     }
 
+    // Bob already owns a leaf at maximum depth in the tree but decides to increase his stake.
+    // Maximum propagation steps thanks to maximum depth.
     #[benchmark]
-    fn join() {
+    fn join_in_place() {
         let alice: T::AccountId = whitelisted_caller();
-        let base_asset = Asset::Ztg;
         let market_id = create_market_and_deploy_pool::<T>(
             alice.clone(),
-            base_asset,
+            Asset::Ztg,
             2u16,
             _10.saturated_into(),
         );
+        let helper = BenchmarkHelper::<T>::new();
+        let bob = helper.accounts().next().unwrap();
+        helper.populate_liquidity_tree_until_full(market_id, bob.clone());
         let pool_shares_amount = _1.saturated_into();
-        let max_amounts_in = vec![u128::MAX.saturated_into(), u128::MAX.saturated_into()];
+        let max_amounts_in = vec![u128::MAX.saturated_into(); 2];
+        // Due to rounding, we need to buy a little more than the pool share amount.
+        let complete_set_amount = _100.saturated_into();
+        helper.set_up_liquidity_benchmark(market_id, bob.clone(), Some(complete_set_amount));
 
-        assert_ok!(T::MultiCurrency::deposit(base_asset, &alice, pool_shares_amount));
-        assert_ok_with_transaction!(T::CompleteSetOperations::buy_complete_set(
+        #[extrinsic_call]
+        NeoSwaps::join(RawOrigin::Signed(bob), market_id, pool_shares_amount, max_amounts_in);
+    }
```

Comment:
I haven't put much thought into this, but I don't see how this could be done. The `assert!` invocations in `BenchmarkHelper` basically ensure that the tree is in the requested state. In the case of `join_in_place`, for example, we know that the tree is full, Bob is already an LP and there are no abandoned nodes. This guarantees that if Bob joins and the extrinsic doesn't fail, then it was a join in place.

On the other hand, I don't think we have any means of dudcing which executing path was taken after the fact.

Reply:
Nevermind, I guess this works? 134c0a2f87bac58e5c1c97bad442135e1924222b